Assets/Storage/DBFile.php

```diff
@@ -528,4 +529,27 @@ public function canViewFile() {
 				->getStore()
 				->canView($this->Filename, $this->Hash);
 	}
+
+	/**
+	 * Generates the URL for this DBFile preview, this is particularly important for images that
+	 * have been manipulated e.g. by {@link ImageManipulation}
+	 * Use the 'updatePreviewLink' extension point to customise the link.
+	 *
+	 * @param null $action
+	 * @return bool|string
+	 */
+	public function PreviewLink($action = null) {
+		// Since AbsoluteURL can whitelist protected assets,
+		// do permission check first
+		if (!$this->failover->canView()) {
```

Comment:
Seems to be a missing check here to see if failover is a file, or even if it's set?